File: go/SpiralMatrix/SpiralMatrix.go
package main

import "fmt"

func spiralMatrix(matrix [][]int) []int {
	array := make([]int, 0)
	spiral(0, len(matrix[0])-1, 0, len(matrix)-1, matrix, &array)
	return array
}

// Time -> 0(M*N) Space ->  0(M*N)
func spiral(startCol, endCol, startRow, endRow int, matrix [][]int, array *[]int) {
	for startRow <= endRow && startCol <= endCol {
		// go right
		right := startCol
		for right <= endCol {
			*array = append(*array, matrix[startRow][right])
			right++
		}

		startRow++
		if startRow > endRow {
			break
		}

		// go down
		down := startRow
		for down <= endRow {
			*array = append(*array, matrix[down][endCol])
			down++
		}

		endCol--
		if endCol < startCol {
			break
		}

		// go left
		left := endCol
		for left >= startCol {
			*array = append(*array, matrix[endRow][left])
			left--
		}

		endRow--
		if startRow > endRow {
			break
		}

		// go up
		up := endRow
		for up >= startRow {
			*array = append(*array, matrix[up][startCol])
			up--
		}
		startCol++
	}
}

func main() {
	fmt.Println(spiralMatrix([][]int{
		{1, 2, 3, 4},
		{5, 6, 7, 8},
		{9, 10, 11, 12},
		{13, 14, 15, 16},
		{17, 18, 19, 20},
		{21, 22, 23, 24},
	}))
}

File: python/spiralMatrix.py
def spiralMatrix(matrix: list[list[int]]) -> list[int]:
    array = []
    spiral(0, len(matrix[0]) - 1, 0, len(matrix) - 1, matrix, array)
    return array


# Time -> 0(M*N) Space ->  0(M*N)
def spiral(
    startCol, endCol, startRow, endRow, matrix: list[list[int]], array: list[int]
):
    while startCol <= endCol and startRow <= endRow:
        # go right
        right = startCol
        while right <= endCol:
            array.append(matrix[startRow][right])
            right += 1

        startRow += 1

        if startRow > endRow:
            break

        # go down
        down = startRow
        while down <= endRow:
            array.append(matrix[down][endCol])
            down += 1

        endCol -= 1

        if startCol > endCol:
            break

        # go left
        left = endCol
        while left >= startCol:
            array.append(matrix[endRow][left])
            left -= 1

        endRow -= 1
        if startRow > endRow:
            break

        # go up
        up = endRow
        while up >= startRow:
            array.append(matrix[up][startCol])
            up -= 1

        startCol += 1


print(
    spiralMatrix(
        [
            [1, 2, 3, 4],
            [5, 6, 7, 8],
            [9, 10, 11, 12],
            [13, 14, 15, 16],
            [17, 18, 19, 20],
            [21, 22, 23, 24],
        ]
    )
)
